File: core/src/test/scala/diplomacy/variant/standard/rule/MovementResolverStandoffSpec.scala
package diplomacy.variant.standard.rule

import diplomacy.UnitSpec
import diplomacy.board._
import diplomacy.rule.OrderResult
import diplomacy.variant.standard.map
import diplomacy.variant.standard.map._
import diplomacy.variant.standard.board.{ Turn => T }
import diplomacy.variant.standard.board._

class MovementResolverStandoffSpec extends UnitSpec {
  val resolver = new MovementResolver() {
    type Turn = T
    type Power = map.Power
  }

  type Executed =
    OrderResult.Executed[resolver.Power, resolver.MilitaryBranch, resolver.Order, resolver.Result]

  "A movement-resolver" should {
    "handle DIAGRAM 4" in {
      val board = new resolver.Board(
        map.map, 1901.Spring - Movement,
        Set(
          DiplomacyUnit(Germany, Army, Ber),
          DiplomacyUnit(Russia, Army, War)
        ),
        Map(), Map(), Map()
      )

      val $ = StandardRuleOrderHelper(board)
      import $._
      val Right(result) = resolver(
        board,
        Set(
          A(Ber).move(Sil),
          A(War).move(Sil)
        )
      )

      result.result should be(Set(
        new Executed(A(Ber).move(Sil), Result.Bounced),
        new Executed(A(War).move(Sil), Result.Bounced)
      ))
      result.board should be(board.copy(
        state = 1901.Spring - Retreat,
        provinceStatuses = Map(Sil.province -> ProvinceStatus.Standoff)
      ))
    }

    "handle DIAGRAM 5" in {
      val board = new resolver.Board(
        map.map, 1901.Spring - Movement,
        Set(
          DiplomacyUnit(Germany, Army, Kie),
          DiplomacyUnit(Germany, Army, Ber),
          DiplomacyUnit(Russia, Army, Pru)
        ),
        Map(), Map(), Map()
      )

      val $ = StandardRuleOrderHelper(board)
      import $._
      val Right(result) = resolver(
        board,
        Set(
          A(Kie).move(Ber),
          A(Ber).move(Pru),
          A(Pru).hold()
        )
      )

      result.result should be(Set(
        new Executed(A(Ber).move(Pru), Result.Bounced),
        new Executed(A(Kie).move(Ber), Result.Bounced),
        new Executed(A(Pru).hold(), Result.Success)
      ))
      result.board should be(board.copy(
        state = 1901.Spring - Retreat
      ))
    }

    "handle DIAGRAM 6" in {
      val board = new resolver.Board(
        map.map, 1901.Spring - Movement,
        Set(
          DiplomacyUnit(Germany, Army, Ber),
          DiplomacyUnit(Russia, Army, Pru)
        ),
        Map(), Map(), Map()
      )

      val $ = StandardRuleOrderHelper(board)
      import $._
      val Right(result) = resolver(
        board,
        Set(
          A(Ber).move(Pru),
          A(Pru).move(Ber))
        )

      result.result should be(Set(
        new Executed(A(Ber).move(Pru), Result.Bounced),
        new Executed(A(Pru).move(Ber), Result.Bounced)
      ))
      result.board should be(board.copy(
        state = 1901.Spring - Retreat
      ))
    }

    "handle DIAGRAM 7" in {
      val board = new resolver.Board(
        map.map, 1901.Spring - Movement,
        Set(
          DiplomacyUnit(England, Fleet, Bel),
          DiplomacyUnit(England, Fleet, Nth),
          DiplomacyUnit(Germany, Army, Hol)
        ),
        Map(), Map(), Map()
      )

      val $ = StandardRuleOrderHelper(board)
      import $._
      val Right(result) = resolver(
        board,
        Set(
          F(Bel).move(Nth),
          F(Nth).move(Hol),
          A(Hol).move(Bel)
        )
      )

      result.result should be(Set(
        new Executed(F(Bel).move(Nth), Result.Success),
        new Executed(F(Nth).move(Hol), Result.Success),
        new Executed(A(Hol).move(Bel), Result.Success)
      ))
      result.board should be(board.copy(
        state = 1901.Spring - Retreat,
        units = Set(
          DiplomacyUnit(England, Fleet, Hol),
          DiplomacyUnit(England, Fleet, Nth),
          DiplomacyUnit(Germany, Army, Bel)
        )
      ))
    }
  }
}

File: test/variant/standard/rule/movement-resolver-standoff-spec.ts
import * as chai from "chai"
import { MovementResolver } from "../../../../src/variant/standard/rule/movement-resolver"
import { StandardRuleHelper as Helper } from "../../../../src/variant/standard/rule/standard-rule-helper"
import { Board, Unit } from "./../../../../src/variant/standard/rule/types"
import { MilitaryBranch } from "./../../../../src/variant/standard/rule/military-branch"
import { State } from "./../../../../src/variant/standard/rule/state"
import { Dislodged } from "./../../../../src/variant/standard/rule/dislodged"
import { Result } from "./../../../../src/variant/standard/rule/result"
import { Phase } from "./../../../../src/variant/standard/rule/phase"
import { ProvinceStatus } from "./../../../../src/variant/standard/rule/province-status"
import { Executed } from "./../../../../src/rule/module"
import { locations as $, Power, map } from "./../../../../src/variant/standard/map/module"
import { Season, Turn } from "./../../../../src/variant/standard/board/module"

const { Army, Fleet } = MilitaryBranch
const { Retreat, Movement } = Phase
const { Spring } = Season

const should = chai.should()

const r = new MovementResolver()

const state1 = new State(new Turn(1901, Spring), Movement)
const state2 = new State(new Turn(1901, Spring), Retreat)

describe("MovementResolver#Standoffs", () => {
  it("DIAGRAM 4", () => {
    const b = new Board<Power>(
      map, state1,
      [new Unit(Army, $.Ber, Power.Germany), new Unit(Army, $.War, Power.Russia)],
      [], []
    )
    const $$ = new Helper(b)
    const { board, results } = r.resolve(
      b, new Set([$$.A($.Ber).move($.Sil), $$.A($.War).move($.Sil)])
    ).result

    Array.from(results).should.have.deep.members([
      new Executed($$.A($.Ber).move($.Sil), Result.Bounced),
      new Executed($$.A($.War).move($.Sil), Result.Bounced)
    ])

    board.should.deep.equal(new Board<Power>(
      map, state2,
      [new Unit(Army, $.Ber, Power.Germany), new Unit(Army, $.War, Power.Russia)],
      [],
      [[$.Sil.province, new ProvinceStatus(null, true)]]
    ))
  })
  it("DIAGRAM 5", () => {
    const b = new Board<Power>(
      map, state1,
      [
        new Unit(Army, $.Kie, Power.Germany), new Unit(Army, $.Ber, Power.Germany),
        new Unit(Army, $.Pru, Power.Russia)
      ],
      [], []
    )
    const $$ = new Helper(b)
    const { board, results } = r.resolve(
      b, new Set([$$.A($.Kie).move($.Ber), $$.A($.Ber).move($.Pru), $$.A($.Pru).hold()])
    ).result

    Array.from(results).should.have.deep.members([
      new Executed($$.A($.Kie).move($.Ber), Result.Bounced),
      new Executed($$.A($.Ber).move($.Pru), Result.Bounced),
      new Executed($$.A($.Pru).hold(), Result.Success)
    ])
    board.should.deep.equal(new Board<Power>(
      map, state2,
      [
        new Unit(Army, $.Kie, Power.Germany), new Unit(Army, $.Ber, Power.Germany),
        new Unit(Army, $.Pru, Power.Russia)
      ],
      [], []
    ))
  })
  it("DIAGRAM 6", () => {
    const b =
      new Board<Power>(
        map, state1,
        [new Unit(Army, $.Ber, Power.Germany), new Unit(Army, $.Pru, Power.Russia)],
        [], []
      )
    const $$ = new Helper(b)
    const { board, results } = r.resolve(
      b, new Set([$$.A($.Ber).move($.Pru), $$.A($.Pru).move($.Ber)])
    ).result

    Array.from(results).should.have.deep.members([
      new Executed($$.A($.Ber).move($.Pru), Result.Bounced),
      new Executed($$.A($.Pru).move($.Ber), Result.Bounced)
    ])
    board.should.deep.equal(new Board<Power>(
      map, state2,
      [new Unit(Army, $.Ber, Power.Germany), new Unit(Army, $.Pru, Power.Russia)],
      [], []
    ))
  })
  it("DIAGRAM 7", () => {
    const b = new Board<Power>(
      map, state1,
      [
        new Unit(Fleet, $.Bel, Power.England), new Unit(Fleet, $.Nth, Power.England),
        new Unit(Army, $.Hol, Power.Germany)
      ],
      [], []
    )
    const $$ = new Helper(b)
    const { board, results } = r.resolve(
      b, new Set([$$.F($.Bel).move($.Nth), $$.F($.Nth).move($.Hol), $$.A($.Hol).move($.Bel)])
    ).result

    Array.from(results).should.have.deep.members([
      new Executed($$.F($.Bel).move($.Nth), Result.Success),
      new Executed($$.F($.Nth).move($.Hol), Result.Success),
      new Executed($$.A($.Hol).move($.Bel), Result.Success)
    ])
    board.should.deep.equal(new Board<Power>(
      map, state2,
      [
        new Unit(Fleet, $.Nth, Power.England), new Unit(Fleet, $.Hol, Power.England),
        new Unit(Army, $.Bel, Power.Germany)
      ],
      [], []
    ))
  })
})
